### PR TITLE
feat: add multi vector support

### PR DESCRIPTION
**Pull request by omriel1**
## Description

This PR introduces multi-vector support!

@efriis @zc277584121 I've moved and modified the PR previously submitted to the main langhcain repo: https://github.com/langchain-ai/langchain/pull/26500

Milvus 2.4 introduced the option for [multi-vector support](https://milvus.io/blog/milvus-2-4-nvidia-cagra-gpu-index-multivector-search-sparse-vector-support.md), which is becoming increasingly popular, especially for use cases like hybrid search (dense + sparse embeddings).

Lately, @ohadeytan  introduced the option to use sparse embeddings in [this PR](https://github.com/langchain-ai/langchain/pull/25284).

Additionally, @zc277584121 already introduced the `MilvusCollectionHybridSearchRetriever`, which enables hybrid search against pre-defined collections directly via pymilvus.

However, this method doesn't take full advantage of the many useful features offered by `langchain_milvus` when building a collection: automatic schema creation, indexing and search parameter creation etc.

This PR intend to make developers life easier, by allowing them to use single-vector or multi-vector with a single `langchain` interface, that create, connect, and search `Milvus`. For example, at IBM and IBM Research this feature is requested by many of our developers and researchers and will be very useful for us. 

## Changes

This PR addresses the limitations described above by introducing the following changes:

1. Allows passing multiple embedding functions with optional matching indexing parameters, search parameters, and vector field names.
2. Dynamically creates the collection using these functions, similar to how it's done for a single embedding function.
3. Adds multiple tests to validate this new feature.

We are eager to have this merged into `langchain-milvus` as we utilize many of langchain features, particularly `langchain-milvus`. We want to continue benefiting from the many valuable features this package provides. We'll make any required changes and will be glad to get any guidance to make it happen!

Twitter handle: @EliyahuOmri, @ohadeytan

### CODE REVIEW

**Comment by codingjaguar** [comment]
Hi @omriel1 thanks for the PR! FYI, the reviewer @zc277584121 is out for a week. Is this PR urgent? Would prefer to wait till the reviewer is back, or we could ask @efriis to take a look first.

File: libs/milvus/tests/integration_tests/vectorstores/test_milvus.py [hunk 1]
```diff
@@ -441,6 +442,106 @@ def test_milvus_vector_field(temp_milvus_db: Any) -> None:
...
+    dense_embedding_func_1 = FakeEmbeddings()
+    dense_embeddings_func_2 = FakeEmbeddings()
+    docsearch = Milvus.from_texts(
+        embedding=[
+            sparse_embedding_func,
+            dense_embedding_func_1,
+            dense_embeddings_func_2,
+        ],
+        texts=fake_texts,
+        connection_args={"uri": "./milvus_demo.db"},
```
Comment (zc277584121) [inline libs/milvus/tests/integration_tests/vectorstores/test_milvus.py:456]: Can this uri be passed in the same way as ut above using temp_milvus_db? Just for the code consistency

File: libs/milvus/tests/integration_tests/vectorstores/test_milvus.py [hunk 2]
```diff
@@ -441,6 +442,106 @@ def test_milvus_vector_field(temp_milvus_db: Any) -> None:
...
+        "metric_type": "IP",
+        "index_type": "HNSW",
+    }
+
+    docsearch = Milvus.from_texts(
+        texts=fake_texts,
+        embedding=[FakeEmbeddings(), FakeEmbeddings()],
+        index_params=[index_param_1, index_param_2],
+        vector_field=["vec_field_1", "vec_field_2"],
+        connection_args={"uri": "./milvus_demo.db"},
```
Comment (zc277584121) [inline libs/milvus/tests/integration_tests/vectorstores/test_milvus.py:479]: ditto

File: libs/milvus/tests/integration_tests/vectorstores/test_milvus.py [hunk 3]
```diff
@@ -441,6 +442,106 @@ def test_milvus_vector_field(temp_milvus_db: Any) -> None:
...
+    embedding_1 = FixedValuesEmbeddings(documents_base_val=0.0, query_val=float(0))
+    # Force the query to always be identical to the embeddings of the *last* document
+    embedding_2 = FixedValuesEmbeddings(
+        documents_base_val=0.0, query_val=float(len(fake_texts))
+    )
+    docsearch = Milvus.from_texts(
+        embedding=[embedding_1, embedding_2],
+        texts=fake_texts,
+        index_params=[index_param_1, index_param_2],
+        connection_args={"uri": "./milvus_demo.db"},
```
Comment (zc277584121) [inline libs/milvus/tests/integration_tests/vectorstores/test_milvus.py:523]: ditto

File: libs/milvus/langchain_milvus/vectorstores/zilliz.py [hunk 4]
```diff
@@ -71,53 +72,74 @@ class Zilliz(Milvus):
...
+                indexes_params: List[dict] = [
+                    {} for _ in range(len(embeddings_functions))
+                ]
+            else:
+                indexes_params = self._as_list(self.index_params)
+
+            for i, embeddings_func in enumerate(embeddings_functions):
+                if not self._get_index(vector_fields[i]):
+                    try:
+                        # If no index params, use a default HNSW based one
```
Comment (zc277584121) [inline libs/milvus/langchain_milvus/vectorstores/zilliz.py:95]: maybe we should keep the comment `If no index params, use a default *AutoIndex* based one` unchanged

File: libs/milvus/langchain_milvus/vectorstores/zilliz.py [hunk 5]
```diff
@@ -71,53 +72,74 @@ class Zilliz(Milvus):
...
+                                }
+
+                        try:
+                            self.col.create_index(
+                                vector_fields[i],
+                                index_params=indexes_params[i],
+                                using=self.alias,
+                            )
+
+                        # If default did not work, most likely on Zilliz Cloud
```
Comment (zc277584121) [inline libs/milvus/langchain_milvus/vectorstores/zilliz.py:117]: ditto

File: libs/milvus/langchain_milvus/vectorstores/zilliz.py [hunk 6]
```diff
@@ -71,53 +72,74 @@ class Zilliz(Milvus):
...
+                        except MilvusException:
+                            # Use AUTOINDEX based index
```
Comment (zc277584121) [inline libs/milvus/langchain_milvus/vectorstores/zilliz.py:119]: ditto

**Comment by zc277584121** [comment]
@omriel1 sorry for delay. thanks for your great contributing. and i have left some small comments here. all others LGTM

File: libs/milvus/langchain_milvus/vectorstores/zilliz.py (hunk 5 above)
Comment (omriel1) [reply libs/milvus/langchain_milvus/vectorstores/zilliz.py:117]: Should I keep "most likely on Zilliz Cloud"?  maybe:
`If default did not work (most likely on Zilliz Cloud), use *HNSW* index instead`
and then I can also remove the next comment (line 119)

**Pull request merged**